parser/src/python.lalrpop

```diff
@@ -759,25 +759,24 @@ ClassPattern: ast::Pattern = {
 }
 
 IfStatement: ast::Stmt = {
-    <location:@L> "if" <test:NamedExpressionTest> ":" <body:Suite> <s2:(<@L> "elif" <NamedExpressionTest> ":" <Suite>)*> <s3:("else" ":" <Suite>)?> => {
-        // Determine last else:
-        let mut last = s3.unwrap_or_default();
-        let end_location = last
+    <location:@L> "if" <test:NamedExpressionTest> ":" <body:Suite> <s2:(<@L> "elif" <NamedExpressionTest> ":" <Suite>)*> <s3:(<@L> "else" ":" <Suite>)?> => {
+        let elif_else_clauses: Vec<_> = s2.into_iter().map(|(start, test, body)| ast::ElifElseClause {
+            range: (start..body.last().unwrap().end()).into(),
+            test: Some(test),
+            body,
+        }).chain(s3.into_iter().map(|(start, body)| ast::ElifElseClause {
+            range: (start..body.last().unwrap().end()).into(),
+            test: None,
+            body,
+        })).collect();
+
+        let end_location = elif_else_clauses
             .last()
-            .or_else(|| s2.last().and_then(|last| last.2.last()))
-            .or_else(|| body.last())
-            .unwrap()
-            .end();
-        // handle elif:
-        for i in s2.into_iter().rev() {
-            let x = ast::Stmt::If(
-                ast::StmtIf { test: Box::new(i.1), body: i.2, orelse: last, range: (i.0..end_location).into() }
-            );
-            last = vec![x];
-        }
+            .map(|last| last.end())
+            .unwrap_or(body.last().unwrap().end());
```

Comment:
May be worth moving this into a callback to not always execute it
```suggestion
            .unwrap_or_else(|| body.last().unwrap().end());
```

Reply:
that will be easier once suite statements make this free